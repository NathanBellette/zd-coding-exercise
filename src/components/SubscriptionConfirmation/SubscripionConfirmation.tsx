import React, {useEffect, useState} from 'react';
import {Subscription} from '../../common/interfaces';
import SubscriptionSummary from '../SubscriptionSummary/SubscriptionSummary';
import SummaryRow from '../SummaryRow/SummaryRow';
import Button from '../Button/Button';
import styles from './SubscriptionConfirmation.module.scss';
import {useHistory} from 'react-router-dom';
import axios from 'axios';
import {useToasts} from "react-toast-notifications";
import Loading from "../Loading/Loading";

export interface ComponentProps {
    previousSubscription: Subscription | undefined;
}

const SubscriptionConfirmation: React.FC<ComponentProps> = ({previousSubscription}) => {
    const [currentSubscription, setCurrentSubscription] = useState<Subscription>();
    const history = useHistory();
    const [loading, setLoading] = useState<boolean>(false);
    const {addToast}  = useToasts();

    useEffect(() => {
        setLoading(true);
        axios.get('/api/current/1')
            .then(response => {
                setLoading(false);
                setCurrentSubscription(response.data.subscription);
            })
            .catch(error => {
                setLoading(false);
                addToast(error.message, {
                    appearance: 'error',
                    autoDismiss: false
                });
            })
            .finally(() => setLoading(false));
    }, []);

    const handleButtonClick = () => {
        history.goBack();
    };

    const planUpdated: boolean = previousSubscription?.plan.id !== currentSubscription?.plan.id;
    const seatsUpdated: boolean = previousSubscription?.seats !== currentSubscription?.seats;
    const priceUpdated: boolean = previousSubscription?.cost !== currentSubscription?.cost;

    return (
        <section>
            <Loading loading={loading} />
            <div className={styles.summaries}>
                <div className={styles.summariesColumn}>
                    <SubscriptionSummary title="Previous Subscription">
                        <SummaryRow
                            label="Plan"
                            updated={planUpdated}
                            value={previousSubscription?.plan.name || ''} />
                        <SummaryRow
                            label="Seats"
                            updated={seatsUpdated}
                            value={previousSubscription?.seats || 0} />
                        <SummaryRow
                            label="Price"
                            updated={priceUpdated}
                            value={previousSubscription?.cost || 0} />
                    </SubscriptionSummary>
                    <div className={styles.buttonRow}>
                        <Button disabled={false} onClick={handleButtonClick}>
                            Back
                        </Button>
                    </div>
                </div>
                <div className={styles.summariesColumn}>
                    <SubscriptionSummary title="Updated Subscription">
                        <SummaryRow
                            label="Plan"
                            updated={planUpdated}
                            value={currentSubscription?.plan.name || ''} />
                        <SummaryRow
                            label="Seats"
                            updated={seatsUpdated}
                            value={currentSubscription?.seats || 0} />
                        <SummaryRow
                            label="Price"
                            updated={priceUpdated}
                            value={currentSubscription?.cost || 0} />
                    </SubscriptionSummary>
                </div>
            </div>
        </section>
    )
};

export default SubscriptionConfirmation;