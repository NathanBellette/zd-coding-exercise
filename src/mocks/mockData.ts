import {Plan, Product, Subscription} from '../interfaces';

export const plans: Plan[] = [
    {
        id: '1',
        name: 'Basic',
        cost: 1
    },
    {
        id: '2',
        name: 'Good',
        cost: 10
    },
    {
        id: '3',
        name: 'Better',
        cost: 100
    },
    {
        id: '4',
        name: 'Best',
        cost: 1000
    }
];

const products: Product[] = [
    {
        id: '1',
        plans: plans
    }
]

export const subscriptions: Subscription[] = [
    {
        id: '1',
        product: products[0],
        plan: plans[0],
        seats: 5,
        cost: 5
    }
];