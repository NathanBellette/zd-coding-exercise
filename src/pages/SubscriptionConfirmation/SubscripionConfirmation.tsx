import React, {useEffect, useState} from 'react';
import {Subscription} from '../../interfaces';
import SubscriptionSummary from '../../components/SubscriptionSummary/SubscriptionSummary';
import SummaryRow from '../../components/SubscriptionSummary/SummaryRow/SummaryRow';
import Button from '../../components/Button/Button';
import styles from './SubscriptionConfirmation.module.scss';
import {useHistory} from 'react-router-dom';
import axios from 'axios';

export interface ComponentProps {
    previousSubscription: Subscription | undefined;
}

const SubscriptionConfirmation: React.FC<ComponentProps> = ({previousSubscription}) => {
    const [currentSubscription, setCurrentSubscription] = useState<Subscription>();
    const history = useHistory();

    useEffect(() => {
        axios.get('/api/current/1')
            .then(response => {
                console.log('current response: ', response.data.subscription);
                setCurrentSubscription(response.data.subscription);
            });
    }, []);

    const handleButtonClick = () => {
        history.goBack();
    };

    const planUpdated: boolean = previousSubscription?.plan.id !== currentSubscription?.plan.id;
    const seatsUpdated: boolean = previousSubscription?.seats !== currentSubscription?.seats;
    const priceUpdated: boolean = previousSubscription?.cost !== currentSubscription?.cost;

    return (
        <section>
            <div className={styles.summaries}>
                <div className={styles.summariesRow}>
                    <SubscriptionSummary title="Previous Subscription">
                        <SummaryRow
                            label="Plan"
                            updated={planUpdated}
                            value={previousSubscription?.plan.name || ''} />
                        <SummaryRow
                            label="Seats"
                            updated={seatsUpdated}
                            value={previousSubscription?.seats || 0} />
                        <SummaryRow
                            label="Price"
                            updated={priceUpdated}
                            value={previousSubscription?.cost || 0} />
                    </SubscriptionSummary>

                    <SubscriptionSummary title="Updated Subscription">
                        <SummaryRow
                            label="Plan"
                            updated={planUpdated}
                            value={currentSubscription?.plan.name || ''} />
                        <SummaryRow
                            label="Seats"
                            updated={seatsUpdated}
                            value={currentSubscription?.seats || 0} />
                        <SummaryRow
                            label="Price"
                            updated={priceUpdated}
                            value={currentSubscription?.cost || 0} />
                    </SubscriptionSummary>
                </div>
                <div className={styles.buttonRow}>
                    <Button disabled={false} onClick={handleButtonClick}>
                        Back
                    </Button>
                </div>
            </div>
        </section>
    )
};

export default SubscriptionConfirmation;