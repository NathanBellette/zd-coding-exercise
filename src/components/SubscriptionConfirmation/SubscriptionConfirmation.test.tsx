import React from 'react';
import {render} from '@testing-library/react';
import SubscripionConfirmation, {ComponentProps} from './SubscripionConfirmation';
import {subscriptions} from '../../mocks/mockData';
import {ToastProvider} from "react-toast-notifications";

describe('<SubscriptionConfirmation />', () => {
    const defaultProps: ComponentProps = {
        previousSubscription: subscriptions[0]
    };

    it('renders the component', () => {
        const container = render(<ToastProvider><SubscripionConfirmation {...defaultProps} /></ToastProvider>);
        expect(container).toMatchSnapshot();
    });
})