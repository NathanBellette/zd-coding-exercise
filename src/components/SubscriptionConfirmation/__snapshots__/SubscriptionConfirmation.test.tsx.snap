// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SubscriptionConfirmation /> renders the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section>
        <div
          class="loading"
        >
          <span
            class="css-1xdhyk6"
          >
            <span
              class="css-1puqkdb"
            />
            <span
              class="css-10y34i5"
            />
            <span
              class="css-1puqkdb"
            />
          </span>
        </div>
        <div
          class="summaries"
        >
          <div
            class="summariesColumn"
          >
            <article
              class="summary"
            >
              <h2>
                Previous Subscription
              </h2>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Plan
                </p>
                <p
                  class="value updated"
                >
                  Basic
                </p>
              </div>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Seats
                </p>
                <p
                  class="value updated"
                >
                  5
                </p>
              </div>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Price
                </p>
                <p
                  class="value updated"
                >
                  5
                </p>
              </div>
            </article>
            <div
              class="buttonRow"
            >
              <button
                class="button"
              >
                Back
              </button>
            </div>
          </div>
          <div
            class="summariesColumn"
          >
            <article
              class="summary"
            >
              <h2>
                Updated Subscription
              </h2>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Plan
                </p>
                <p
                  class="value updated"
                />
              </div>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Seats
                </p>
                <p
                  class="value updated"
                >
                  0
                </p>
              </div>
              <div
                class="row"
              >
                <p
                  class="label"
                >
                  Price
                </p>
                <p
                  class="value updated"
                >
                  0
                </p>
              </div>
            </article>
          </div>
        </div>
      </section>
    </div>
    <div
      class="react-toast-notifications__container css-19n335a-ToastContainer"
    />
  </body>,
  "container": <div>
    <section>
      <div
        class="loading"
      >
        <span
          class="css-1xdhyk6"
        >
          <span
            class="css-1puqkdb"
          />
          <span
            class="css-10y34i5"
          />
          <span
            class="css-1puqkdb"
          />
        </span>
      </div>
      <div
        class="summaries"
      >
        <div
          class="summariesColumn"
        >
          <article
            class="summary"
          >
            <h2>
              Previous Subscription
            </h2>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Plan
              </p>
              <p
                class="value updated"
              >
                Basic
              </p>
            </div>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Seats
              </p>
              <p
                class="value updated"
              >
                5
              </p>
            </div>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Price
              </p>
              <p
                class="value updated"
              >
                5
              </p>
            </div>
          </article>
          <div
            class="buttonRow"
          >
            <button
              class="button"
            >
              Back
            </button>
          </div>
        </div>
        <div
          class="summariesColumn"
        >
          <article
            class="summary"
          >
            <h2>
              Updated Subscription
            </h2>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Plan
              </p>
              <p
                class="value updated"
              />
            </div>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Seats
              </p>
              <p
                class="value updated"
              >
                0
              </p>
            </div>
            <div
              class="row"
            >
              <p
                class="label"
              >
                Price
              </p>
              <p
                class="value updated"
              >
                0
              </p>
            </div>
          </article>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
